[docs] AWS database discovery (#30913)
gravitational/teleport: docs/pages/database-access/guides/aws-discovery.mdx

---
title: AWS Database Auto-Discovery
description: How to configure Teleport to discover AWS-hosted databases.
---

(!docs/pages/includes/database-access/discovery-guide-brief.mdx cloud="AWS" !)

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An AWS account with permissions to create and attach IAM policies.
- One or more database servers hosted by AWS database services.
- A host to run the Teleport Discovery Service.

## Step 1/4. Generate a join token

> The Discovery Service require a valid join token to connect to the cluster.

Reviewer: ```suggestion
The Discovery Service requires a valid join token to connect to the cluster.
```

Generate a join token by running the following command against your Teleport
Auth Service and save it in `/tmp/token` on the host that will run the
Discovery Service.

```code
$ tctl tokens add --type=discovery
```

(!docs/pages/includes/database-access/alternative-methods-join.mdx!)

## Step 2/4. Configure the Discovery service

Enabling AWS database discovery requires that the `discovery_service.aws`
section includes at least one entry and that `discovery_service.aws.types`
includes one of database types listed in the sample YAML below. 

> Now create the `teleport.yaml` file on the host:

Reviewer: ```suggestion
Create a `teleport.yaml` file similar to the following on the host that will run the Discovery Service:
```

    # 'opensearch' - discovers AWS OpenSearch Redis databases.
  - types: ["rds"]
    regions: ["us-east-1"]
    tags:
      "env": "prod" # Match database resource tags where tag:env=prod

  - types: ["redshift", "redshift-serverless"]
    regions: ["us-west-1"]
    tags:
      "env": "prod"
    # Optional AWS role that the Discovery Service will assume to discover
    # AWS-hosted databases. The IAM identity assigned to the host must be able
    # to assume this role.
    assume_role_arn: "arn:aws:iam::123456789012:role/example-role-name"
    # Optional AWS external ID that the Database Service will use to assume
    # a role in an external AWS account.
    external_id: "example-external-id"
```

Adjust the keys under `discovery_service.aws` to match your AWS databases.

(!docs/pages/kubernetes-access/discovery/includes/discovery-group.mdx!)

## Step 3/4. Bootstrap IAM permissions

Create an IAM role and attach it to the host that will run the Discovery
Service.

Teleport can bootstrap IAM permissions for the Discovery Service using the
`teleport discovery bootstrap` command. You can use this command in automatic
or manual mode:

> - In automatic mode, Teleport will attempt to create appropriate IAM policies

Reviewer: ```suggestion
- In automatic mode, Teleport attempts to create the appropriate IAM policies
```

  and attach them to the specified IAM roles. This requires IAM permissions to
  create and attach IAM policies.

> - In manual mode, Teleport will print required IAM policies. You can then create

Reviewer: ```suggestion
- In manual mode, Teleport prints the required IAM policies. You can then create
```

  and attach them manually using the AWS management console.

<Tabs>
  <TabItem label="Automatic">
  Either temporarily give IAM admin permissions to the host of the Discovery
  Service or copy the service YAML configuration file to your desktop where you
  have the IAM admin permissions.

  Use this command to bootstrap the permissions automatically with YAML
  configuration file of the Discovery Service:

  ```code
  $ teleport discovery bootstrap \
    --attach-to-role arn:aws:iam::<Var name="aws-account-id"/>:role/<Var name="iam-role-name"/> \
    --policy-name TeleportDatabaseDiscovery \
    -c <Var name="/etc/teleport.yaml"/>
  ```
  </TabItem>

  <TabItem label="Manual">

>   Use this command to display required IAM policies which you will then create
>   in your AWS console:

Reviewer: ```suggestion
  Use the following command to display the required IAM policies that you need to create
  in your AWS console:
```

  ```code
  $ teleport discovery bootstrap --manual \
    --attach-to-role arn:aws:iam::<Var name="aws-account-id"/>:role/<Var name="iam-role-name"/> \
    --policy-name TeleportDatabaseDiscovery \
    -c <Var name="/etc/teleport.yaml"/>
  ```
  </TabItem>
</Tabs>

<Details title="Bootstrapping with assume_role_arn in config">
When `assume_role_arn` is configured for AWS matchers, `teleport discovery

> bootstrap` command will determine permissions required for the bootstrap target
> AWS IAM identity using the following logic:

Reviewer: ```suggestion
bootstrap` command determines the permissions required for the bootstrap target
AWS IAM identity using the following logic:
```

- When the target does not match `assume_role_arn` in any AWS matcher in the
  configuration file, the target is assumed to be the Teleport Discovery
  Service's AWS IAM identity and permissions are bootstrapped for all the AWS
  matchers without `assume_role_arn`.
- When an `--attach-to-role` target matches an `assume_role_arn` setting for

>   AWS matchers in the configuration file, permissions will be bootstrapped only

Reviewer: ```suggestion
  AWS matchers in the configuration file, permissions are bootstrapped only
```

  for those AWS matchers.

You will need to run the bootstrap command once with the Teleport Discovery
Service's IAM identity as the policy attachment target, and once for each AWS
IAM role that is used for `assume_role_arn`.
</Details>

## Step 3/3. Start the Discovery Service

(!docs/pages/includes/start-teleport.mdx service="the Discovery Service"!)

> Once the Discovery Service starts, database servers matching the tags and
> regions specified in the AWS section will be added to the Teleport cluster
> automatically.

Reviewer: ```suggestion
After the Discovery Service starts, database servers matching the tags and
regions specified in the AWS section are added to the Teleport cluster
automatically.
```

---
title: AWS Database Auto-Discovery
description: How to configure Teleport to discover AWS-hosted databases.
---

(!docs/pages/includes/database-access/discovery-guide-brief.mdx cloud="AWS" !)

> <Admonition type="tip">

Reviewer 2: I would incorporate this information into the Prerequisites, since otherwise it strays from the focus of this guide a bit.

Maybe we could recommend either one host for both the Discovery Service and the Database Service or separate hosts for each.

Author: I will just remove the tip. I think the current `Prerequisites` implies a separate host for discovery service anyway, which should be the preferred configuration.

> Discovery Service can run independently on a different machine from Database
> Services.

Reviewer 2: ```suggestion
The Discovery Service can run independently on a different machine from Database
Service instances.
```

For example, you can run an instance of the Discovery Service for discovering
databases across all your AWS accounts, and an instance of the Database Service
in each VPC for accessing database servers in private subnets.

The same Discovery Service can also be used for Azure database discovery,
server instance discovery and Kubernetes discovery.

</Admonition>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An AWS account with permissions to create and attach IAM policies.
- One or more database servers hosted by AWS database services.
- A host to run the Teleport Discovery Service.

## Step 1/4. Generate a join token

> The Discovery Service require a valid auth token to connect to the cluster.

Reviewer 2: I think we should replace "auth token" with "join token" to be consistent with the heading

bootstrap` command will determine permissions required for the bootstrap target
AWS IAM identity using the following logic:

- When the target does not match `assume_role_arn` in any AWS matcher in the
  configuration file, the target is assumed to be the Teleport Discovery
  Service's AWS IAM identity and permissions are bootstrapped for all the AWS
  matchers without `assume_role_arn`.
- When an `--attach-to-role` target matches an `assume_role_arn` setting for
  AWS matchers in the configuration file, permissions will be bootstrapped only
  for those AWS matchers.

You will need to run the bootstrap command once with the Teleport Discovery
Service's IAM identity as the policy attachment target, and once for each AWS
IAM role that is used for `assume_role_arn`.
</Details>

## Step 3/3. Start the Discovery Service

(!docs/pages/includes/start-teleport.mdx service="the Discovery Service"!)

Once the Discovery Service starts, database servers matching the tags and
regions specified in the AWS section will be added to the Teleport cluster
automatically.

You can list them with the `tctl get db` command or check a specific database
with `tctl get db/<database-name>`.

## Next

> - Learn [Dynamic Registration](./dynamic-registration.mdx) by Database Service.

Reviewer 2: ```suggestion
- Learn about [Dynamic Registration](./dynamic-registration.mdx) by the Teleport Database Service.
```
